File: Rasbet/Domain/BetDomain/Bet.cs
namespace Domain;

public abstract class Bet
{
    public int Id { get; set; }
    private double amount { get; set; }
    private double wonValue { get; set; }

    public DateTime Start { get; set; }
    public DateTime ?End { get; set; }

    // utilizador que realizou a aposta
    public string UserId { get; set; }

    public BetState State { get; set; } = BetState.Open;

    public double Amount
    {
        get { return amount; }
        set { amount = Math.Max(0.1, value); }
    }

    public double WonValue
    {
        get { return wonValue; }
        set { wonValue = Math.Max(0, value); }
    }

    protected Bet()
    {
    }

    public Bet(double amount, DateTime start, string userId)
    {
        Amount = amount;
        Start = start;
        UserId = userId;
    }

    public abstract void SetFinishBet(int betTypeId, List<int> odds);
    public void FinishBet(bool won)
    {
        State = won ? BetState.Won : BetState.Lost;
    }
}

File: rasbet_app/src/app/models/bet.ts
import { IBetType } from "./betType";
import { CollectiveGame, IGame } from "./game";
import { IOdd } from "./odd";

export type betState = "Open" | "Won" | "Lost";

export interface IBet {
  amount: number;
  wonValue: number;
  start: Date;
  end?: Date;
  state: betState;
  user: string;
  selections: ISelectionGame[];
  odd: number
}

export interface ISelectionGame {
  odd: number;
  game: IGame;
  win: boolean;
}

export interface ISelection {
  betType: IBetType;
  oddValue: number;
  odd: IOdd;
  game: CollectiveGame;
}

export interface IBetDetails {
  amount: number;
  userId: string;
}

export interface ISimpleDetails extends IBetDetails {
  selection: ISelection;
}

export interface IMultipleDetails extends IBetDetails {
  selections: ISelection[];
}

export interface ICreateSelection {
  betTypeId: number;
  oddId: number;
  odd: number;
  gameId: number;
}

export interface ICreateBet {
  amount: number;
  userId: string;
}

export interface ICreateBetSimple extends ICreateBet {
  selection: ICreateSelection;
}

export interface ICreateBetMultiple extends ICreateBet {
  selections: ICreateSelection[];
}
